File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxi=Integer.MIN_VALUE;
    int helper(TreeNode root){
        if(root==null){
            return 0;
        }
        int lbs=helper(root.left);
        int rbs=helper(root.right);
        maxi=Math.max(maxi, lbs+root.val+rbs);

        return Math.max(0,Math.max(lbs+root.val,rbs+root.val));

    }
    public int maxPathSum(TreeNode root) {
        if(root==null){
            return 0;
        }
        helper(root);
        return maxi;
    }
}

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxi=INT_MIN;
    int helper(TreeNode* root){
        if(root==NULL){
            return 0;
        }
        int l=helper(root->left);
        int r=helper(root->right);

        maxi=max({maxi,root->val,root->val+l,root->val+r,root->val+l+r});

        return max({root->val+l,root->val+r,root->val});
    }
    int maxPathSum(TreeNode* root) {
         helper(root);
         return maxi;
    }
};